src/java.base/share/classes/java/util/regex/MatchResult.java

```diff
@@ -184,6 +304,53 @@ public interface MatchResult {
      *
      * @return The number of capturing groups in this matcher's pattern
      */
-    public int groupCount();
+    int groupCount();
+
+    /**
+     * Returns an unmodifiable map from capturing group names to group numbers.
+     * If there are no named groups, returns an empty map.
+     *
+     * @return an unmodifiable map from capturing group names to group numbers
+     *
+     * @implSpec
+     * The default implementation of this method throws
+     * {@link UnsupportedOperationException}.
+     *
+     * @apiNote
+     * This method must be overridden by an implementation that supports named groups.
+     *
+     * @since 20
+     */
+    default Map<String,Integer> namedGroups() {
+        throw new UnsupportedOperationException("namedGroups()");
+    }
+
+    private int groupIndex(String name) {
+        Objects.requireNonNull(name, "Group name");
+        Integer index = namedGroups().get(name);
+        if (index != null) {
+            return index;
+        }
+        throw new IllegalArgumentException("No group with name <" + name + ">");
+    }
+
+    /**
+     * Returns whether {@code this} contains a valid match from
+     * a previous match or find operation.
+     *
+     * @return whether {@code this} contains a valid match
+     *
+     * @implSpec
+     * The default implementation of this method throws
+     * {@link UnsupportedOperationException}.
+     *
+     * @apiNote
+     * This method must be overridden by an implementation.
```

Comment:
This is a bit odd. It sounds like existing MatchResult implementations (outside the JDK) are now invalid. I think it really means something like, "This method must be overridden by an implementation in order to provide valid information about whether this MatchResult contains a match." I'm not sure whether saying this is necessary; it could be omitted.

Probably also needs an `@throws UnsupportedOperationException` in case the match information is unavailable.